Whitespace before a build instruction causes the parser to panic from an empty stack.

Whitespace before a build instruction causes the parser to panic from an empty stack.

AFAIK leading whitespace shouldn't be unparseable (please correct me if I'm wrong); but if so, we should return a more appropriate error message, rather than panicking and throwing out a full backtrace.

`earthly version v0.6.30 743687991d227f2f7e2e6ba11a90f443b4f99c67 linux/amd64; Ubuntu 20.04.5 LTS (Focal Fossa)`

### MRPE

#### Okay:
```Earthfile
VERSION --use-cache-command 0.6

## Build Arguments ###########################################################

ARG SourceImage=ubuntu:focal
#^^^
 
RUN true
```
---
```shell
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as podman container (earthly-buildkitd)

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
Error: build target: build main: failed to solve: Earthfile line 0:0 target build not found
in              +build
```

### Crashes:
```Earthfile
VERSION --use-cache-command 0.6

## Build Arguments ###########################################################

 ARG SourceImage=ubuntu:focal
#^^^
 
RUN true
```
---
```shell
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as podman container (earthly-buildkitd)

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

panic:  [recovered]
        panic: Empty Stack [recovered]
        panic: Empty Stack [recovered]
        panic: Empty Stack

goroutine 168 [running]:
github.com/earthly/earthly/ast/parser.(*EarthParser).EarthFile.func2()
        /earthly/ast/parser/earth_parser.go:722 +0xfe
panic({0x1069f20, 0x1434ab0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmts.func2()
        /earthly/ast/parser/earth_parser.go:1855 +0xfe
panic({0x1069f20, 0x1434ab0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/antlr/antlr4/runtime/Go/antlr.(*BaseLexer).PopMode(0xc0003d2980?)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/lexer.go:267 +0x129
github.com/earthly/earthly/ast.(*lexer).popRecipeMode(0xc0003d2980)
        /earthly/ast/lexer.go:58 +0x9d
github.com/earthly/earthly/ast.(*lexer).processIndentation(0xc0003d2980, {0x144d0e0, 0xc00033de80})
        /earthly/ast/lexer.go:160 +0x293
github.com/earthly/earthly/ast.(*lexer).NextToken(0xc0003d2980)
        /earthly/ast/lexer.go:104 +0x1c5
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).fetch(0xc0000eb280, 0x3e8)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:128 +0x69
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).Fill(0xc0000eb280)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:370 +0x3d
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).GetTextFromInterval(0xc0000eb280, 0xc000823728)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:334 +0x3b
github.com/antlr/antlr4/runtime/Go/antlr.(*CommonTokenStream).GetTextFromTokens(0x417c6d?, {0x144d0e0?, 0xc00033d7d0?}, {0x144d0e0, 0xc00033d7d0})
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/common_token_stream.go:325 +0x87
github.com/antlr/antlr4/runtime/Go/antlr.(*DefaultErrorStrategy).ReportNoViableAlternative(0x2?, {0x1457800, 0xc0006cb5b0}, 0xc000a453b0)
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/error_strategy.go:258 +0xa2
github.com/antlr/antlr4/runtime/Go/antlr.(*DefaultErrorStrategy).ReportError(0x0?, {0x1457800?, 0xc0006cb5b0?}, {0x143f120?, 0xc000a453b0?})
        /go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200225173536-225249fdaef5/runtime/Go/antlr/error_strategy.go:132 +0xc6
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmt.func2()
        /earthly/ast/parser/earth_parser.go:2065 +0xbb
panic({0x1130580, 0xc000a453b0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmt(0xc0006cb5b0)
        /earthly/ast/parser/earth_parser.go:2127 +0x4f9
github.com/earthly/earthly/ast/parser.(*EarthParser).Stmts(0xc0006cb5b0)
        /earthly/ast/parser/earth_parser.go:1877 +0x1f7
github.com/earthly/earthly/ast/parser.(*EarthParser).EarthFile(0xc0006cb5b0)
        /earthly/ast/parser/earth_parser.go:767 +0x385
github.com/earthly/earthly/ast.newEarthfileTree({0xc0003180a0, 0x9}, 0xc0000ca520, {0x1449800?, 0xc0000ea200})
        /earthly/ast/ast.go:83 +0x348
github.com/earthly/earthly/ast.Parse({0x1443d58?, 0xc0000d6190}, {0xc0003180a0, 0x9}, 0x1)
        /earthly/ast/ast.go:25 +0x14c
github.com/earthly/earthly/buildcontext.(*Resolver).parseEarthfile.func1({0x1443d58?, 0xc0000d6190?}, {0x1069f20?, 0xc0002c6480?})
        /earthly/buildcontext/resolver.go:117 +0xa5
github.com/earthly/earthly/util/syncutil/synccache.(*SyncCache).Do.func1()
        /earthly/util/syncutil/synccache/synccache.go:44 +0x50
created by github.com/earthly/earthly/util/syncutil/synccache.(*SyncCache).Do
        /earthly/util/syncutil/synccache/synccache.go:40 +0x151
```

--- a/ast/lexer.go
+++ b/ast/lexer.go
@@ -47,6 +47,12 @@ func (l *lexer) getMode() int {
 // popRecipeMode removes the recipe mode stack frame and then places back anything that existed
 // on top of it.
 func (l *lexer) popRecipeMode() {
+	defer func() {
+		// Recovering here prevents a panic due to erroneous indentation. The
+		// tokens will still fail to match our parser grammar, so we have no
+		// need to add an error of our own.
+		_ = recover()
+	}()
 	m := l.getMode()
 	if m == parser.EarthLexerRECIPE {
 		// Special case: nothing above.

--- a/tests/invalid/Earthfile
+++ b/tests/invalid/Earthfile
@@ -20,18 +20,24 @@ WORKDIR /test
 test-trailing-backslash:
     DO +RUN_EARTHLY_ARGS --earthfile=trailing-backslash.earth --should_fail=true --target=+base
 
+test-leading-whitespace:
+    DO +RUN_EARTHLY_ARGS --earthfile=leading-whitespace.earth --should_fail=true --target=+base --output_contains="syntax error"
+
 test-all:
     BUILD +test-trailing-backslash
+    BUILD +test-leading-whitespace
 
 RUN_EARTHLY_ARGS:
     COMMAND
     ARG earthfile
     ARG target
     ARG should_fail=false
+    ARG output_contains
     DO tests+RUN_EARTHLY \
         --earthfile=$earthfile \
         --target=$target \
         --should_fail=$should_fail \
+        --output_contains=$output_contains \
         --DOCKERHUB_AUTH=$DOCKERHUB_AUTH \
         --DOCKERHUB_USER_SECRET=$DOCKERHUB_USER_SECRET \
         --DOCKERHUB_TOKEN_SECRET=$DOCKERHUB_TOKEN_SECRET \

--- a/tests/invalid/leading-whitespace.earth
+++ b/tests/invalid/leading-whitespace.earth
@@ -1,0 +1,9 @@
+VERSION 0.7
+FROM alpine:3.15
+
+ ARG foo
+# ^^ leading whitespace at the top level causes INDENT/DEDENT confusion in the
+# parser. This error should be caught and output, but should not cause earthly
+# to panic.
+
+RUN true

